### PR TITLE
[SPARK-14716][SQL] Added support for partitioning in FileStreamSink

### PR DESCRIPTION
**Pull request by tdas**
# What changes were proposed in this pull request?

Support partitioning in the file stream sink. This is implemented using a new, but simpler code path for writing parquet files - both unpartitioned and partitioned. This new code path does not use Output Committers, as we will eventually write the file names to the metadata log for "committing" them. 

This patch duplicates < 100 LOC from the WriterContainer. But its far simpler that WriterContainer as it does not involve output committing. In addition, it introduces the new APIs in FileFormat and OutputWriterFactory in an attempt to simplify the APIs (not have Job in the `FileFormat` API, not have bucket and other stuff in the `OutputWriterFactory.newInstance()` ).
# Tests
- New unit tests to test the FileStreamSinkWriter for partitioned and unpartitioned files
- New unit test to partially test the FileStreamSink for partitioned files (does not test recovery of partition column data, as that requires change in the StreamFileCatalog, future PR).
- Updated FileStressSuite to test number of records read from partitioned output files.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 1]
```diff
@@ -76,6 +76,8 @@ private[sql] object FileSourceStrategy extends Strategy with Logging {
 
       val partitionColumns =
         l.resolve(files.partitionSchema, files.sqlContext.sessionState.analyzer.resolver)
+      logInfo(s"Partition columns: $partitionColumns")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: not needed

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala [hunk 2]
```diff
@@ -371,6 +382,97 @@ private[sql] class DefaultSource
...
+      SQLConf.PARQUET_WRITE_LEGACY_FORMAT.key,
+      sqlConf.writeLegacyParquetFormat.toString)
+
+    // Sets compression scheme
+    conf.set(ParquetOutputFormat.COMPRESSION, parquetOptions.compressionCodec)
+
+    conf
+  }
+
+  private val serializableConf = new SerializableConfiguration(preparedHadoopConf)
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: We should broadcast this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala [hunk 3]
```diff
@@ -40,17 +48,22 @@ object FileStreamSink {
 class FileStreamSink(
     sqlContext: SQLContext,
     path: String,
-    fileFormat: FileFormat) extends Sink with Logging {
+    fileFormat: FileFormat,
+    partitionColumnNames: Seq[String],
+    options: Map[String, String]
+  ) extends Sink with Logging {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala]: Nit: we don't normally wrap the `)`

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 4]
```diff
@@ -347,6 +347,14 @@ abstract class OutputWriterFactory extends Serializable {
       bucketId: Option[Int], // TODO: This doesn't belong here...
       dataSchema: StructType,
       context: TaskAttemptContext): OutputWriter
+
+  /**
+   * Returns a new instance of [[OutputWriter]] that will write data to the given path.
+   * @since 2.0.0
+   */
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Can we clarify the difference between these (i.e. which one is old and which is new?).

**Comment by marmbrus** [comment]
General structure looks good!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala (hunk 2 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: Isnt the whole the closure broadcasted? This whole output factory will be in the closure and therefore broadcasted.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRelation.scala]: You know, I thought that was the case but then I found an example where closure capture of the value for a broadcast slowed down the jobs by an order of magnitude.  So i'm not convinced this is safe.  I guess we should either verify or broadcast.

**Comment by tdas** [comment]
@marmbrus Updated with tests and more docs.

**Comment by tdas** [comment]
@marmbrus Please take a look.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 5]
```diff
@@ -477,6 +491,20 @@ trait FileFormat {
   }
 
   /**
+   * Returns a [[OutputWriterFactory]] for generating output writers that can write data.
+   * This method is current used only by FileStreamSinkWriter to generate output writers that
+   * does not use output committers to write data. The OutputWriter generated by the returned
+   * [[OutputWriterFactory]] must implement the method `newInstance(path)`..
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: nit: `newWriter(path)`

**Comment by marmbrus** [comment]
LGTM

**Pull request closed without merging**